### PR TITLE
Appending to TSDB: Move crucial actions to defer

### PR DESCRIPTION
**Pull request by beorn7**
With defer having less of a performance penalty, there is no reason
not to do those crucial operations via defer.

Context: With isolation in place, if we forget to Commit/Rollback, the
low watermark will get stuck forever.

The current code should not have any bugs, but moving to defer helps
to avoid future bugs.

This is also moving the `closeAppend` in the `Commit` implementation
itself to defer. If logging to the WAL fails, we would have missed the
`closeAppend`.

Related to #6909 .

### CODE REVIEW

**Comment by roidelapluie** [comment]
I put my release shepherd veto on this pull request until we use go1.14, which we are not at this moment.

**Comment by beorn7** [comment]
Yes, sure.

For context: https://github.com/prometheus/golang-builder/pull/85 describes what is holding us back.

**Comment by roidelapluie** [comment]
> Yes, sure.
> 
> For context: [prometheus/golang-builder#85](https://github.com/prometheus/golang-builder/pull/85) describes what is holding us back.

Yes, sorry, github does not send updated comments by email.

File: scrape/scrape.go [hunk 1]
```diff
@@ -1258,67 +1259,71 @@ const (
...
+			return
+		}
+		err = app.Commit()
+	}()
 
-	if err := sl.addReportSample(app, scrapeHealthMetricName, ts, health); err != nil {
-		app.Rollback()
-		return err
+	if err = sl.addReportSample(app, scrapeHealthMetricName, ts, health); err != nil {
+		return
```
Comment (brian-brazil) [inline scrape/scrape.go]: It's a little weird to see returning without the err, could we make this explicit?

File: tsdb/head.go [hunk 2]
```diff
@@ -1068,13 +1069,14 @@ func (a *headAppender) Commit() error {
...
 	a.head.metrics.samplesAppended.Add(float64(total))
 	a.head.updateMinMaxTime(a.mint, a.maxt)
-	a.head.iso.closeAppend(a.appendID)
 
 	return nil
 }
 
 func (a *headAppender) Rollback() error {
-	a.head.metrics.activeAppenders.Dec()
+	defer a.head.metrics.activeAppenders.Dec()
```
Comment (brian-brazil) [inline tsdb/head.go]: Why defer rather than immediately?

File: scrape/scrape.go [hunk 3]
```diff
@@ -1258,67 +1259,71 @@ const (
...
 
 	var health float64
-	if err == nil {
+	if scrapeErr == nil {
 		health = 1
 	}
 	app := sl.appender()
+	defer func() {
+		if err != nil {
+			app.Rollback()
```
Comment (roidelapluie) [inline scrape/scrape.go]: while we are at it should we use multierr ?

File: scrape/scrape.go (hunk 1 above)
Comment (beorn7) [reply scrape/scrape.go]: While I avoid naked returns in general, in this particular case it is a legitimate pattern (inspecting and changing the returned error in `defer`). And In that case, I'd like to embrace the naked return completely and mark it clearly by, well, a naked return.

Comment (beorn7) [reply scrape/scrape.go]: Having said that, if we have a style convention to still put the names in the return statement, I'm happy to do so, but I'm not aware of any.

File: tsdb/head.go (hunk 2 above)
Comment (beorn7) [reply tsdb/head.go]: To be consistent with what we do in `Commit`. I also think it is clearer to do it in the end in this way. Makes debugging easier if for example one of the locks cannot get acquired.

File: scrape/scrape.go (hunk 3 above)
Comment (beorn7) [reply scrape/scrape.go]: I'd like to keep unrelated changes out of this PR.

Comment (beorn7) [reply scrape/scrape.go]: Or perhaps not completely unrelated, depending on how you see it.

But still, let's do one (non-trivial) thing per PR.

**Comment by beorn7** [comment]
I have rebased this on top of the release-2.17 branch and changed the base.

Ironically, this doesn't really help with the panic in #6977 but I think, now that we have seen panics getting recovered because they are in the HTTP serving path, we should be conservative and merge this into release-2.17.

I'll keep checking the querying path for similar requirements to move to defer.

**Comment by roidelapluie** [comment]
Can we reach consensus regarding the explicit returns discussion? @brian-brazil @beorn7

**Comment by roidelapluie** [comment]
> Ironically, this doesn't really help with the panic in #6977

Can we do something for that kind of panics?

**Comment by beorn7** [comment]
> Can we do something for that kind of panics?

Done in #6983 . (o:

**Comment by brian-brazil** [comment]
> Can we reach consensus regarding the explicit returns discussion?

I'm not that pushed on it.

**Comment by roidelapluie** [comment]
Thanks!

**Pull request merged**